### ISSUE TITLE
how to figure out  legacy?

### ISSUE THREAD

**Issue by xuan-xuan6** [ultralytics/ultralytics#19561]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

"
   if m is C3k2:  # for M/L/X sizes
                legacy = False
                if scale in "mlx":
                    args[3] = True
"
Hello, author. Currently, I am in the process of porting the C3k2 module of YOLOv11 to the previous YOLOv5 (anchor-based). When I was reading this part, I didn't understand the meanings of "legacy" and "# for M/L/X sizes". What impact does it have on the C3k2 module whether "legacy" is True or False? Additionally, does "# for M/L/X sizes" mean that the C3k2 module can only be used in models of M, L, and X sizes, and not in those of N and S sizes?
I'm really looking forward to your reply.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @xuan-xuan6, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Documentation](https://docs.ultralytics.com/) for helpful resources where many usage examples and module explanations are provided.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a ❓ Question about custom implementations like porting modules, we kindly ask you to share further details. For example:
- Your intended goal with the `C3k2` module and its integration.
- Any related code snippets such as the modified YOLOv5 code.
- Training or inference results if available.

Also, please verify you're using the latest `ultralytics` package to ensure compatibility and access to the latest features. You can upgrade to the latest version with:

```bash
pip install -U ultralytics
```

For a deeper understanding of the `C3k2` module and its configurations (including `legacy` and size-specific behaviors), it might help to trace its functionality or inspect relevant usages in the YOLOv11 codebase. Consider also exploring our [Model Training Tips](https://docs.ultralytics.com/guides/model-training-tips/) for more insights.

### Join the Community

Looking to connect with others working on similar projects? Engage with the Ultralytics community:
- For quick discussions, join us on [Discord](https://discord.com/invite/ultralytics) 🎧
- For in-depth questions, check out [Discourse](https://community.ultralytics.com/)
- Join lively debates and feature requests on our [Subreddit](https://reddit.com/r/Ultralytics)

### Run YOLO Anywhere

If you're testing YOLO implementations, you can explore verified environments with preinstalled dependencies (Python, PyTorch, etc.):
- **Notebooks**: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Cloud Platforms**: See our [GCP](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/) or [AWS](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/) quickstart guides.
- **Docker Image**: See the [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

### CI Status

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

This is an automated response to guide you in the right direction. An Ultralytics engineer will review your question and provide further assistance soon 😊

**Comment by xuan-xuan6** [comment]
> Legacy is used in Detect head which is different for YOLO11 vs. previous YOLO.
> 
> [ultralytics/ultralytics/nn/modules/head.py](https://github.com/ultralytics/ultralytics/blob/23a90142dc66fbb180fe1bb513a2adc44322c978/ultralytics/nn/modules/head.py#L48)
> 
> Line 48 in [23a9014](/ultralytics/ultralytics/commit/23a90142dc66fbb180fe1bb513a2adc44322c978)
> 
>  if self.legacy

Thank you for your reply. If I want to port the C3k2 and C2PSA modules to the original YOLOv5 (anchor-based), do I need to use the "legacy" option? does "# for M/L/X sizes" mean that the C3k2 module can only be used in models of M, L, and X sizes, and not in those of N and S sizes?

**Comment by Y-T-G** [comment]
You don't need legacy. 

The scale just means C3k is enabled by default for those scales. But for n and m, C3k is also used, but you need to enable explicitly in config.

https://github.com/ultralytics/ultralytics/blob/23a90142dc66fbb180fe1bb513a2adc44322c978/ultralytics/cfg/models/11/yolo11-seg.yaml#L26

**Comment by xuan-xuan6** [comment]
# YOLOv5 🚀 by Ultralytics, AGPL-3.0 license



> You don't need legacy.
> 
> The scale just means C3k is enabled by default for those scales. But for n and s, C3k is also used, but you need to enable explicitly in config.
> 
> [ultralytics/ultralytics/cfg/models/11/yolo11-seg.yaml](https://github.com/ultralytics/ultralytics/blob/23a90142dc66fbb180fe1bb513a2adc44322c978/ultralytics/cfg/models/11/yolo11-seg.yaml#L26)
> 
> Line 26 in [23a9014](/ultralytics/ultralytics/commit/23a90142dc66fbb180fe1bb513a2adc44322c978)
> 
>  - [-1, 2, C3k2, [512, True]]

# Parameters
nc: 3 # number of classes
depth_multiple: 0.50 # model depth multiple
width_multiple: 0.50 # layer channel multiple 1.25 -> 0.50
anchors:
  - [10, 13, 16, 30, 33, 23] # P3/8
  - [30, 61, 62, 45, 59, 119] # P4/16
  - [116, 90, 156, 198, 373, 326] # P5/32

# YOLOv5 v6.0 backbone
backbone:
  # [from, number, module, args]
  [
    [-1, 1, Conv, [64, 3, 2]], # 0-P1/2
    [-1, 1, Conv, [128, 3, 2]], # 1-P2/4
    [-1, 2, C3k2, [256, True, 0.25]],
    [-1, 1, Conv, [256, 3, 2]], # 3-P3/8
    [-1, 2, C3k2, [512, False, 0.25]],
    [-1, 1, Conv, [512, 3, 2]], # 5-P4/16
    [-1, 2, C3k2, [512, True]],
    [-1, 1, Conv, [1024, 3, 2]], # 7-P5/32
    [-1, 2, C3k2, [1024, True]],
    [-1, 1, SPPF, [1024, 5]], # 9
    [-1, 2, C2PSA, [1024]], #10
  ]

# YOLOv5 v6.0 head
head: [
    [-1, 1, nn.Upsample, [None, 2, "nearest"]],
    [[-1, 6], 1, Concat, [1]], # cat backbone P4
    [-1, 2, C3k2, [512, False]], # 13

    [-1, 1, nn.Upsample, [None, 2, "nearest"]],
    [[-1, 4], 1, Concat, [1]], # cat backbone P3
    [-1, 2, C3k2, [256, False]], # 16 (P3/8-small)

    [-1, 1, Conv, [256, 3, 2]],
    [[-1, 13], 1, Concat, [1]], # cat head P4
    [-1, 2, C3k2, [512, False]], # 19 (P4/16-medium)

    [-1, 1, Conv, [512, 3, 2]],
    [[-1, 10], 1, Concat, [1]], # cat head P5
    [-1, 2, C3k2, [1024, True]], # 22 (P5/32-large)

    [[16, 19, 22], 1, Segment, [nc, anchors, 32, 256]], # Detect(P3, P4, P5)
  ]


When I printed the new structure in YOLOv5 and compared it with the names of each layer of the original YOLOv11s-seg model, I found there were significant differences. I haven't found out where the problem lies.



![Image](https://github.com/user-attachments/assets/25f5978b-a9ab-4507-b6fe-be4d3cd97eee)


my save model name code as below（in origin anchor based yolov5）：

from models.yolo import Model
# load YOLOv5 model
yolov5_cfg = 'models/segment/model_s_c3k2_c2psa_seg_2.yaml'  
yolov5_model = Model(yolov5_cfg)
yolov5_state_dict = yolov5_model.state_dict()


with open('yolov5_layer_names.txt', 'w') as f:
    for name, param in yolov5_state_dict.items():
        f.write(name + '\n')


my save model code as below（in this repo yolov11s-seg.pt）：

from ultralytics import YOLO
model = YOLO("yolo11l-seg.pt", task="segment")

yolov11_state_dict = model.state_dict()

with open('yolov11s-seg_layer_names.txt', 'w') as f:
    for name, param in yolov11_state_dict.items():
        f.write(name + '\n')

**Comment by xuan-xuan6** [comment]
> You don't need legacy.
> 
> The scale just means C3k is enabled by default for those scales. But for n and s, C3k is also used, but you need to enable explicitly in config.
> 
> [ultralytics/ultralytics/cfg/models/11/yolo11-seg.yaml](https://github.com/ultralytics/ultralytics/blob/23a90142dc66fbb180fe1bb513a2adc44322c978/ultralytics/cfg/models/11/yolo11-seg.yaml#L26)
> 
> Line 26 in [23a9014](/ultralytics/ultralytics/commit/23a90142dc66fbb180fe1bb513a2adc44322c978)
> 
>  - [-1, 2, C3k2, [512, True]]

is model = YOLO("yolo11s-seg.pt", task="segment")

**Comment by glenn-jocher** [comment]
Glad you resolved the issue! For future reference, the architectural differences stem from YOLO11's anchor-free design versus YOLOv5's anchor-based approach, which impacts layer naming and structure. When combining modules between versions, ensure compatibility between detection heads (Segment in your case) and backbone architectures. Let us know if you need clarification on specific architectural implementations.
